gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -0,0 +1,160 @@
+                "WEEK_FORWARD": {
+                    "bought_kWh": 32.432775153437625,
+                    "sold_kWh": 32.432775153437625,
+                },
+                "DAY_FORWARD": {
+                    "bought_kWh": 6.486555030687522,
+                    "sold_kWh": 6.486555030687522,
+                },
+            },
+            DateTime(2020, 2, 1, 0, 0, 0): {...},
+            ...
+            DateTime(2020, 12, 1, 0, 0, 0): {...},
+        },
+    }
+    """
+    def __init__(
+            self, asset_uuid: str, asset_peak_kWh: float,
+            resolution: AggregationResolution):
+
+        self.asset_uuid = asset_uuid
+        self.asset_peak_kWh = asset_peak_kWh
+        self.resolution = resolution
+
+        self._asset_volume_time_series_buffer: Dict[DateTime, Dict] = {}
+        self._trade_profile_generator = ForwardTradeProfileGenerator(self.asset_peak_kWh)
+
+    def add(self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add asset time series to asset volume time series."""
+
+        self._add_total_energy_bought(asset_stats, market_type)
+        self._add_total_energy_sold(asset_stats, market_type)
+
+    def save_time_series(self):
+        """Save asset volume time series in the DB."""
+        # TODO: should be implemented.
+        raise NotImplementedError()
+
+    def _add_total_energy_bought(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of bought energy."""
+        energy_kWh = asset_stats.total_energy_bought
+        if energy_kWh <= 0:
+            return
+        profile = self._trade_profile_generator.generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type)
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "bought_kWh")
+
+    def _add_total_energy_sold(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of sold energy."""
+        energy_kWh = asset_stats.total_energy_sold
+        if energy_kWh <= 0:
+            return
+        profile = self._trade_profile_generator.generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type
+        )
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "sold_kWh")
+
+    def _add_to_volume_time_series(
+            self, time_slot: DateTime, value: float,
+            market_type: AvailableMarketTypes, attribute_name: str):
+        """Add time slot value to the correct time slot to asset volume time series."""
+        time_slot = self._adapt_time_slot(time_slot)
+        volume_time_series = self._get_asset_volume_time_series(time_slot.start_of("year"))
+        volume_time_series[time_slot][market_type.name][attribute_name] += value
```

Comment:
As discussed offline, the time_slot here has a different meaning. It is the key that divides the data into yearly chunks. So I suggest to rename it here into e.g. `year_selector` or`year_key` or similar.
Also, i think there is a bug here, because also in Line 107 you need to select the start of the year, right?
```suggestion
        time_slot = self._adapt_time_slot(time_slot)
        year_key = time_slot.start_of("year")
        volume_time_series = self._get_asset_volume_time_series(year_key)
        volume_time_series[year_key][market_type.name][attribute_name] += value
```

Reply:
Thanks. I renamed it to `year`:
```python
    def _get_asset_volume_time_series(self, year: DateTime):
```
Also, it's not a bug as the function returns the requested volume time series (as the name `_get_...` suggests).

Reply:
Umm, it's returning the time series actually: At the end of the function ...

Reply:
Is github failing me again?
I do not see a return statement. 
<img width="610" alt="Screenshot 2022-10-04 at 10 58 10" src="https://user-images.githubusercontent.com/19326608/193778724-8c5d35ce-f6f8-40e1-8af8-28bfa85e182f.png">

Reply:
The return statement is inside `_get_asset_volume_time_series`. After time series is returned, we assign it to `volume_time_series` and then modify it.

Reply:
Yeah, but the `self._asset_volume_time_series_buffer` is not adapted here AFAIU.
IMO, the 
``` python
self._asset_volume_time_series_buffer[year] = time_series
```
should be performed after
```python
volume_time_series[str(time_slot)][product_type.name][attribute_name] += value
```
Even if this works because there is a reference, to the `_asset_volume_time_series_buffer` I suggest to work with return values. At some point we decided to not use references if possible but to return and assign. 
I hope this makes sense

Reply:
So, what I mean is:
``` python
def _add_to_volume_time_series(
        self, time_slot: DateTime, value: float,
        product_type: AvailableMarketTypes, attribute_name: str):
    """Add time slot value to the correct time slot to asset volume time series."""
    time_slot = self._adapt_time_slot(time_slot)
    year = time_slot.start_of("year")
    volume_time_series = self._get_asset_volume_time_series(year=year)
    volume_time_series[str(time_slot)][product_type.name][attribute_name] += value
    self._asset_volume_time_series_buffer[year] = volume_time_series
```

Reply:
Fixed.